text/0625-helper-managers.md

```diff
@@ -0,0 +1,504 @@
+```
+
+```hbs
+{{title "My Document Title"}}
+```
+
+Addons such as [ember-page-title](https://github.com/adopted-ember-addons/ember-page-title)
+use this to make helpers that can be added to the template to specify
+_app behavior_ declaratively. This is a much better way to approach certain
+types of behavior and APIs, compared to the alternative of using mixins and
+lifecycle hooks to manage them. Its benefits include:
+
+1. Behavior can be _self-contained_. Some APIs need to run at multiple points in
+   based on a component's lifecycle, such as for a plugin that needs to be setup
+   on initialization and torn down upon destruction. Using lifecycle hooks for
+   this forces users to split their API across multiple touch points in a
+   component, which requires a lot of boilerplate and can make it difficult to
+   understand how the whole system works together.
+
+2. It doesn't require _multiple inheritance_. Mixins and strategies like them
+   create complicated inheritance hierarchies that can be difficult to debug.
+   A side-effecting helper does not insert itself into the inheritance
+   hierarchy, it is a child of the template instead, which is much easier to
+   reason about in practice.
+
+3. They are highly _composable_. Helpers, like components, can be used multiple
+   times in a template and can be used within `{{if}}` and `{{each}}` blocks.
+   Combined with their ability to be hold a self contained lifecycle, this makes
+   them a powerful tool for composing declarative behavior.
+
+4. They can be _destroyed_. They tie in naturally to the destruction APIs that
+   we have recently added to Ember, and that allows their lifecycle to be
+   managed in a self contained way.
+
+However, this pattern has some issues today, mostly stemming from the fact that
+they execute _during_ render, which is not an ideal time for side-effecting.
+They can also be abused to modify app state, which can lead to difficult to
+follow code paths reminiscent of observers. These issues stem from a mismatch
+between two different goals, the goal of calculating a result or value, and the
+goal of triggering side-effects.
+
+The `isScheduledEffect` capability would schedule side-effecting helpers to execute
+_after_ render, and would _disable_ Ember's state mutations while they were
+running. This would ensure that side-effecting helpers run at the optimal time,
+and do not enable antipatterns and complicated codepaths.
+
+## Detailed design
+
+This RFC proposes adding the `setHelperManager` and `capabilities` APIs,
+imported from `@ember/helper`. Like `setComponentManager` and
+`setModifierManager`, `setHelperManager` receives a callback that is passed
+the owner, and returns an instance of the helper manager. When a helper
+definition is resolved by Ember, it will look up the manager recursively on the
+definition's prototype chain until it finds a helper manager. If it does not
+find one, it will throw an error.
+
+```ts
+type HelperDefinition = object;
```

Comment:
You can thank JS for this one. It’s wild, but because you can do this—

```js
function foo() {
  return 'JavaScript is wild';
}
foo.bar = true;
```

—what else is TS going to do? :sob:

Reply:
FWIW, that doesn't seem very bizarre to me 😸, and in fact `function() {} instanceof Object` is true (which also supports what TS is doing). But `typeof function() {}` is **not** `object` so `typeof` and `instanceof` disagree a bit.

Either way, IMHO, it doesn't _hurt_ to say `object | Function` as far as I can tell.

Reply:
Yeah, the interplay between `Object`, `object`, `typeof`, and `Function` in TS is admittedly a bit funky. `object` ≠ `typeof something === 'object'`, and there’s no equivalent to `'function'` from typeof (or most of the others). 😭